Code/Framework/AzCore/AzCore/EBus/EBusSharedDispatchMutex.h

```diff
@@ -0,0 +1,187 @@
+        {
+            m_eventMutex.lock();
+        }
+
+        void EventMutexUnlockExclusive()
+        {
+            m_eventMutex.unlock();
+        }
+
+        void EventMutexLockShared()
+        {
+            m_eventMutex.lock_shared();
+        }
+
+        void EventMutexUnlockShared()
+        {
+            m_eventMutex.unlock_shared();
+        }
+
+    private:
+        AZStd::shared_mutex m_eventMutex;
+        AZStd::mutex m_callstackMutex;
+
+        // This custom mutex type should only be used with the lock guards below since it needs additional context to know which
+        // mutex to lock and what type of lock to request. If you get a compile error due to these methods being private, the EBus
+        // declaration is likely missing one or more of the lock guards below.
+        void lock(){}
+        void unlock(){}
+    };
+
+    // Custom lock guard to handle Connection lock management.
+    // This will lock/unlock the event mutex with an exclusive lock. It will assert and disallow
+    // exclusive locks if currently inside of a shared lock.
+    template<class EBusType>
+    class EBusSharedDispatchMutexConnectLockGuard
+    {
+    public:
+        AZ_FORCE_INLINE EBusSharedDispatchMutexConnectLockGuard(EBusSharedDispatchMutex& mutex, AZStd::adopt_lock_t)
+            : m_mutex(mutex)
+        {
+        }
+
+        AZ_FORCE_INLINE explicit EBusSharedDispatchMutexConnectLockGuard(EBusSharedDispatchMutex& mutex)
+            : m_mutex(mutex)
+        {
+            AZ_Assert(!EBusType::IsInDispatchThisThread(), "Can't connect/disconnect while inside an event dispatch.");
+            m_mutex.EventMutexLockExclusive();
+        }
+        AZ_FORCE_INLINE ~EBusSharedDispatchMutexConnectLockGuard()
+        {
+            m_mutex.EventMutexUnlockExclusive();
+        }
+
+    private:
+        AZ_FORCE_INLINE EBusSharedDispatchMutexConnectLockGuard(EBusSharedDispatchMutexConnectLockGuard const&) = delete;
+        AZ_FORCE_INLINE EBusSharedDispatchMutexConnectLockGuard& operator=(EBusSharedDispatchMutexConnectLockGuard const&) = delete;
+        EBusSharedDispatchMutex& m_mutex;
+    };
+
+    // Custom lock guard to handle Dispatch lock management.
+    // This will lock/unlock the event mutex with a shared lock. It also allows for recursive shared locks by only holding the
+    // shared lock at the top level of the recursion.
+    template<class EBusType>
+    class EBusSharedDispatchMutexDispatchLockGuard
+    {
+    public:
+        AZ_FORCE_INLINE EBusSharedDispatchMutexDispatchLockGuard(EBusSharedDispatchMutex& mutex, AZStd::adopt_lock_t)
+            : m_mutex(mutex)
+        {
+        }
+
+        AZ_FORCE_INLINE explicit EBusSharedDispatchMutexDispatchLockGuard(EBusSharedDispatchMutex& mutex)
+            : m_mutex(mutex)
+        {
+            if (!EBusType::IsInDispatchThisThread())
+            {
+                m_ownSharedLockOnThread = true;
+                m_mutex.EventMutexLockShared();
```

Comment:
What is this boolean for?
It is not thread safe either or a thread local.

So I am trying to figure out the purpose of this variable.

Reply:
The boolean is to track whether or not this specific lock guard is the one that acquired the shared_lock, so that when the lock guard is destroyed, it knows that it should release the shared_lock.

Lock guards are created on the stack and aren't accessed from multiple threads, so they don't need thread safety on their local variables.

Here's how this works (I'll add comments explaining):
- Each thread that has an EBus call ends up creating a lock guard.
- The lock guard checks (via EBusType::IsInDispatchThisThread) whether or not it is the first EBus call to occur on this thread. 
- If it is, it sets the boolean and share locks the shared_mutex. If not, it doesn't, and accepts that something higher up the callstack has a share lock on the shared_mutex.
- If a recursive call to the EBus occurs on the thread, it does _not_ grab the share lock.

Reply:
Added comments.